calico/common.py

```diff
@@ -435,34 +451,68 @@ def validate_endpoint(config, combined_id, endpoint):
         raise ValidationFailed(" ".join(issues))
 
 
-def validate_rules(profile_id, rules):
+def validate_tier_data(tier, data):
+    issues = []
+    if not VALID_ID_RE.match(tier):
+        issues.append("Invalid profile_id '%r'." % tier)
+
+    if not isinstance(data, dict):
+        raise ValidationFailed("Expected tier data to be a dict.")
+
+    if "order" in data:
+        order = data["order"]
+        if not isinstance(order, numbers.Number):
+            issues.append("Tier data contained non-numeric order field")
+    else:
+        issues.append("Missing 'order' field in tier data")
+
+    if issues:
+        raise ValidationFailed(" ".join(issues))
+
+
+def validate_profile(profile_id, profile, require_selector=False,
+                     require_order=False):
     """
-    Ensures that the supplied rules are valid. Once this routine has returned
-    successfully, we know that all required fields are present and have valid
-    values.
+    Validates and normalises the given rules dictionary.
 
-    :param profile_id: Profile ID from etcd
-    :param rules: rules list as read from etcd
+    As side effects to the input dict:
+
+    * Fields that are set to None are removed completely.
+    * Selectors are replaced with SelectorExpression objects.  Parsing now
+      ensures that the selectors are valid and ensures that equal selectors
+      compare and hash equally.  For example: "a == 'b'" and "a=='b'" are
+      different strings that parse to the same selector.
+
+    Once this routine has returned successfully, we know that all
+    required fields are present and have valid values.
+
+    :param str|TieredPolicyId profile_id: Profile ID from etcd
+    :param dict profile: profile dict as read from etcd
```

Comment:
I'll split this into two separate functions that re-use a validate_rules function.  that should be clearer.